### DISCUSSION TITLE
Query upcoming jobs

### DISCUSSION THREAD

**Discussion by bencyoung-Fignum** [quartznet/quartznet#1235] [Q&A]
I'd like to be able to create a REST API to serve a UI for a scheduler service we're basing off Quartz.NET. Things I'd like to be able to show are:

1) Upcoming jobs in order of due date
2) Recently run jobs
etc

One option is to materialize this into our own tables using a job listener and things like GetNextFireTime. That has the nice property that it's separate from Quartz and we're not interfering but I can't help thinking it would just be duplicating exactly what Quartz already knows. At the same time I can't see a way of efficiently doing this via the API itself.

Has anyone else done similar things?

**Comment by ghost** [comment]
This is not an answer. I'm just piggy-backing on what Ben wrote. This is a major requirement from our business as well:  They want to be able to see what jobs are running today, tomorrow, this week, for the month, etc. If we use Quartz we will write a web ui and web api to allow users to schedule jobs. If I can't query quartz for upcoming jobs, then we'd also have to store job information in our own tables, but doing that run the risk of the two sources becoming out of sync.

**Comment by lahma** [comment]
You can query the database and create the view you need. Just make sure that you have no transaction level that would lock anything, basically dirty reads, otherwise Quartz will be affected.

You can get the next fire time always from DB, that's what gets updated and maintained. Calculating of next one after fire can include custom logic so there won't ever be "the next after next" info available.

For recently run/history you need to use the history plugin and adapt to your needs. I would say that usual case is to create your own one with custom logic, cleanup schedules for such history tables also vary by scenario.

**Reply by ghost** [reply]
Thanks much for the reply. Yes, I had seen the next_run_time field in the database and the logic to update it in Quartz code. Unfortunately our business wants to look at days, weeks, months into the future to see what jobs will be running. I do have the job and trigger history plugins working. I log that information to custom database tables and it would be helpful.

Another proof of concept I'm working on to design our own scheduler. I've modeled the schedule and job database schema similar to SQL Agent (with sql server) where we use fields like frequency_type, frequency_interval, frequency_subinterval, repetition_interval and repetition_duration. With those settings, I have written sql to tell me what jobs will run at any specified minute (which the Scheduler component would query to find out what it should run) and also sql to tell me what jobs will run during a range of time (so the business knows)...  With cron expressions, that gets to be a bit difficult...

Thanks again. As I have said, your scheduler is really amazing.

**Reply by lahma** [reply]
Yes, you could query such things against Quartz tables and even get the future schedules with some SQL/stored proc black magic *given* that a) only simple triggers b) no calendars c) disregard misfire policies causing possible multiple runs.

But you can get the next fire time as a report, then if you "open the trigger in the UI" you could load the trigger information and then use `TriggerUtils.ComputeFireTimes*` helpers to get list of next expected fire times. This would be calculated dynamically and when requested, no state in DB for those.

> Thanks again. As I have said, your scheduler is really amazing.

Thanks! ❤️
